File: test-case/char/chars2d/chars2d.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass
class Solution:
    def toUpper(self, matrix: List[List[str]]) -> List[List[str]]:
        diff = ord('a') - ord('A')
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                matrix[i][j] = chr(ord(matrix[i][j]) - diff)
        return matrix

if __name__ == '__main__':
    init_logging()
    from soda.unittest import Validators
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().toUpper)
    # OR use struct tester
    # work = TestWork.forStruct(CLASS)

    # step [3]: setup other options
    # work.validator = (e,r) => bool
    work.compareSerial = True
    work.run()

File: test-case/char/chars2d/chars2d.go
package main

import (
    . "missingjs.com/soda/leetcode"
    "missingjs.com/soda/unittest"
    "missingjs.com/soda/util"
)

var logger = util.Logger
var _used ListNode

func toUpper(matrix [][]byte) [][]byte {
    diff := 'a' - 'A'
    for i := 0; i < len(matrix); i++ {
        for j := 0; j < len(matrix[i]); j++ {
            matrix[i][j] -= byte(diff)
        }
    }
    return matrix
}

func main() {
    // create tester by function
    work := unittest.CreateWork(toUpper)
    // OR create by struct factory
    // work := unittest.CreateWorkForStruct(Constructor)
    // work.SetValidator(func(e,r)bool)
    work.CompareSerial = true
    work.Run()
}
